pkg/daemon/update.go

```diff
@@ -1644,9 +1646,86 @@ func (dn *Daemon) updateSSHKeys(newUsers []ign3types.PasswdUser) error {
 	return nil
 }
 
+// Gets the labels attached to the OS image and uses that to infer the OS version.
+func (dn *Daemon) getNewOSVersion(imgURL string) (osr osrelease.OperatingSystem, err error) {
+	glog.Infof("Looking up image labels for new OS %q to infer new OS version", imgURL)
+
+	// The image inspection code is copy / pasted from the Rebase() method in rpm-ostree.go.
+	var imageData *types.ImageInspectInfo
+	if imageData, _, err = imageInspect(imgURL); err != nil {
+		if err != nil {
+			var podmanImgData *imageInspection
+			glog.Infof("Falling back to using podman inspect")
+			if podmanImgData, err = podmanInspect(imgURL); err != nil {
+				return
+			}
+			defer exec.Command("podman", "rmi", imgURL).Run()
+
+			osr, err = osrelease.InferFromOSImageLabels(podmanImgData.Labels)
+			return
+		}
+	}
+
+	osr, err = osrelease.InferFromOSImageLabels(imageData.Labels)
+	return
+}
+
+// Returns true if we're updating from RHCOS8 -> RHCOS9. Returns false for all
+// other cases.
+func (dn *Daemon) isRHCOS9Upgrade(newImageOS osRelease) bool {
+	return dn.os.IsEL() && !dn.os.IsEL9() && newImageOS.IsEL() && newImageOS.IsEL9()
+}
+
+// Prepares for an upgrade from RHCOS 8 to RHCOS 9 by doing the following:
+// - Installing a newer version of rpm-ostree from the extensions container.
+// - Restarting rpm-ostree.service.
+//
+// Assumes:
+// - We've called ExtractExtensionsImage
+// - We've called addExtensionsRepo or addLayeredExtensionsRepo.
+func (dn *Daemon) prepForRHCOS9(newImageURL string) error {
+	// Gets the image labels for the new OS image and infers what the OS version
+	// is based upon their contents.
+	newOSInfo, err := dn.getNewOSVersion(newImageURL)
+	if err != nil {
+		return fmt.Errorf("could not get new OS version info: %w", err)
+	}
+
+	// If we don't have an RHCOS 8 -> RHCOS 9 upgrade, return early.
+	if !dn.isRHCOS9Upgrade(newOSInfo) {
+		return nil
+	}
+
+	glog.Infof("Detected RHCOS 8 -> RHCOS 9 upgrade, upgrading rpm-ostree first")
+
+	// Ensures that the extensions repo is available first.
+	if _, err := os.Stat(extensionsRepo); err != nil {
+		return fmt.Errorf("extensions repo not found: %w", err)
+	}
+
+	// Upgrades rpm-ostree.
+	// TODO(zzlotnik): Is this the right command to upgrade rpm-ostree?
+	if _, err := runGetOut("rpm", "-Uvh", "rpm-ostree"); err != nil {
```

Comment:
Is this the right way to upgrade rpm-ostree with the extensions container repo set up?

Reply:
One tricky bit is we'll likely need to run this code in the *host* mount namespace.  So something like:

```
systemd-run -qP /bin/sh -c 'rpm-ostree usroverlay; rpm -Uvh /path/to/extensions/hotfixes/8/*.rpm && systemctl restart rpm-ostreed'
```

or so.

The actual path to the mounted hotfixes will be a `hotfixes` subdirectory of the extensions path.

Reply:
Gotcha. Yeah, that may take a bit to get right or I'll have to drop it in a temp file first.